workers/typescript/src/workflows/kitchen_sink.ts

```diff
@@ -0,0 +1,295 @@
+  SearchAttributes
+} from '@temporalio/common';
+import Long from 'long';
+import WorkflowInput = temporal.omes.kitchen_sink.WorkflowInput;
+import WorkflowState = temporal.omes.kitchen_sink.WorkflowState;
+import Payload = temporal.api.common.v1.Payload;
+import DoSignalActions = temporal.omes.kitchen_sink.DoSignal.DoSignalActions;
+import IActionSet = temporal.omes.kitchen_sink.IActionSet;
+import DoActionsUpdate = temporal.omes.kitchen_sink.DoActionsUpdate;
+import IAction = temporal.omes.kitchen_sink.IAction;
+import IPayload = temporal.api.common.v1.IPayload;
+import IAwaitableChoice = temporal.omes.kitchen_sink.IAwaitableChoice;
+import IExecuteActivityAction = temporal.omes.kitchen_sink.IExecuteActivityAction;
+import ActivityCancellationType = temporal.omes.kitchen_sink.ActivityCancellationType;
+import IDuration = google.protobuf.IDuration;
+import IWorkflowState = temporal.omes.kitchen_sink.IWorkflowState;
+
+const reportStateQuery = defineQuery<IWorkflowState, [Payload]>('report_state');
+const actionsSignal = defineSignal<[DoSignalActions]>('do_actions_signal');
+const actionsUpdate = defineUpdate<IPayload | undefined, [DoActionsUpdate]>('do_actions_update');
+
+export async function kitchenSink(input: WorkflowInput | undefined): Promise<IPayload | undefined> {
+  let workflowState: IWorkflowState = WorkflowState.create();
+  const actionsQueue = new Array<IActionSet>();
+
+  async function handleActionSet(actions: IActionSet): Promise<IPayload | null> {
+    let rval: IPayload | null = null;
+
+    if (!actions.concurrent) {
+      for (const action of actions.actions ?? []) {
+        const actionRval = await handleAction(action);
+        if (actionRval) {
+          rval = actionRval;
+        }
+      }
+      return rval;
+    }
+
+    // Concurrent actions run concurrently but we return early if any return a value
+    const promises = new Array<Promise<void>>();
+    for (const action of actions.actions ?? []) {
+      promises.push(handleAction(action).then((actionRval) => {
+        if (actionRval) {
+          rval = actionRval;
+        }
+      }));
+    }
+    const allComplete = Promise.all(promises);
+    await Promise.any([allComplete, condition(() => rval !== undefined)]);
+
+    return rval;
+  }
+
+  async function handleAction(action: IAction): Promise<IPayload | null | undefined> {
+    // console.log('Handling an action', action);
+    async function handleAwaitableChoice<PR extends Promise<PRR>, PRR>(
+      promise: PR, // TODO: should maybe be factory
+      choice: IAwaitableChoice | null | undefined,
+      afterStarted: (_: Promise<PRR | void>) => Promise<void> = async (_) => {
+        await sleep(1);
+      },
+      afterCompleted: (_: Promise<PRR | void>) => Promise<void> = async (task) => {
+        await task;
+      }
+    ) {
+      const cancelScope = new CancellationScope();
+      let didCancel = false;
+
+      const cancellablePromise = cancelScope.run(() => promise).catch((err) => {
```

Comment:
The promise should be created in the cancellation scope for it to be attached to it, otherwise the command that the promise represents will not be cancelled.